Get started with bcdata vignette (#88)
bcgov/bcdata: vignettes/bcdata.Rmd

---
title: "Get started with bcdata"
author: "" 
date: ""
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{bcdata}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warning = FALSE,
  message = FALSE
)

library(bcdata)
library(ggplot2)
```

The `bcdata` [R](https://www.r-project.org/) package contains functions for searching & retrieving data from the [B.C. Data Catalogue]( https://catalogue.data.gov.bc.ca).

> The [B.C. Data Catalogue](https://www2.gov.bc.ca/gov/content?id=79B5224167334667A44C9E8B5143D0C5) is the place to find British Columbia Government data, applications and web services. Much of the data are released under the [Open Government Licence --- British Columbia](https://www2.gov.bc.ca/gov/content/data/open-data/open-government-licence-bc), as well as numerous other [licences](https://catalogue.data.gov.bc.ca/dataset?download_audience=Public).

Reviewer: ```suggestion
The [B.C. Data Catalogue](https://www2.gov.bc.ca/gov/content?id=79B5224167334667A44C9E8B5143D0C5) is the place to find British Columbia Government data, applications, and web services. Much of the data are released under the [Open Government Licence --- British Columbia](https://www2.gov.bc.ca/gov/content/data/open-data/open-government-licence-bc), as well as numerous other [licences](https://catalogue.data.gov.bc.ca/dataset?download_audience=Public).
```

Reviewer: Sorry, couldn't resist the Oxford comma

Author: 🤣 The sentence came straight from the [program web page](https://www2.gov.bc.ca/gov/content/data/bc-data-catalogue/bc-data-catalogue-about), which is following the [style guide](https://www2.gov.bc.ca/gov/content/governments/services-for-government/policies-procedures/web-content-development-guides/writing-for-the-web/web-style-guide/punctuation)---which falls into the _no_ Oxford Camp. This is a ⛰ for sure.

Reviewer: Argh! Ok, I will have to concede then...

  
  
You can install `bcdata` directly from GitHub using the [remotes](https://cran.r-project.org/package=remotes) package:

```{r, eval=FALSE}
install.packages("remotes")

remotes::install_github("bcgov/bcdata")
library(bcdata)
```

### `bcdc_browse()`

`bcdata::bcdc_browse()` let's you access the B.C. Data Catalogue web interface directly from R---opening the home page in your default browser:

```{r, eval=FALSE}
## Take me to the B.C. Data Catalogue home page
bcdc_browse()
```

If you know the catalogue record "human-readable" name or permanent ID you can open directly to the record web page:

```{r, eval=FALSE}
## Take me to the B.C. Winery Locations catalogue record using the record name
bcdc_browse("bc-winery-locations") 

## Take me to the B.C. Winery Locations catalogue record using the record ID
bcdc_browse("1d21922b-ec4f-42e5-8f6b-bf320a286157")
```

> 

Reviewer: Would we want a screenshot of the catalogue page here?

You can also use `bcdc_browse()` to open to search results in the catalogue:

```{r, eval=FALSE}
## Take me to the catalogue search results for 'wineries'
bcdc_browse("wineries")
```

### `bcdc_search()`

`bcdata::bcdc_search()` let's you search records in the B.C. Data Catalogue, returning the search results in your R session. 

Let's search the catalogue for records that contain the word "recycling":

```{r}
## Give me the catalogue search results for 'recycling'
bcdc_search("recycling")
```

> You can set the number of records to be returned from the search and/or you can customize your search using the catalogue search facets `license_id`, `download_audience`, `type`, `res_format`, `sector`, and `organization`:

Reviewer: ```suggestion
You can set the number of records to be returned from the search and/or you can customize your search using the catalogue search _facets_ `license_id`, `download_audience`, `type`, `res_format`, `sector`, and `organization`:
```

```{r}
## Give me the first catalogue search result for 'recycling'
bcdc_search("recycling", n = 1)

## Give me the catalogue search results for 'recycling' where the 
## data is tabular and licenced under Open Government Licence – British Columbia
bcdc_search("recycling", type = "Dataset", license_id = "2")
```

You can see all valid values for the catalogue search facets using `bcdata::bcdc_search_facets()`:

```{r}
## Valid values for search facet 'type'
bcdc_search_facets(facet = "type")

> ```

Reviewer: I wonder if an additional example with licenses would be helpful here, especially since licenses are specified by number?

Author: Or maybe just change to the licence example? I'll try it out.

```{r}
## Get the txt data resource from the `bc-schools-programs-offered-in-schools`
## catalogue record
bcdc_get_data("b1f27d1c-244a-410e-a361-931fac62a524", resource = 'a393f8cf-51ec-42c6-8449-4cea4c75385c')
```

`bcdc_get_data()` will also detect if the data resource is a geospatial file, and automatically reads and returns it in as an [`sf` object](https://r-spatial.github.io/sf/) in your R session. Let's get the air zones for British Columbia:

```{r}
bcdc_search("air zones", res_format = "geojson")

bc_az_metadata <- bcdc_get_record("e8eeefc4-2826-47bc-8430-85703d328516")

## Get the Air Zone geospatial data ()
bc_az <- bcdc_get_data(bc_az_metadata, resource = "c495d082-b586-4df0-9e06-bd6b66a8acd9")

## Look at the air zone geospatial data with ggplot()
bc_az %>% 
  ggplot() +
  geom_sf()
```

**Note:** The `bcdata` package supports downloading _most_ file types, however the package does _not_ retrieve .zip or compressed resource files at this time.

### `bcdc_query_geodata()`

> Many geospatial files in the B.C. Data Catalogue are available as a [Web Map or Web Feature Service (WMS/WFS)](https://www2.gov.bc.ca/gov/content?id=95D78D544B244F34B89223EF069DF74E). While `bcdc_get_data()` will retrieve the WMS/WFS geospatial automatically from the web for you, sometimes the geospatial file is very large--and slow to download---and you may only want _some_ of the data. `bcdc_query_geodata()` let's you get & query catalogue geospatial data available as WMS/WFS, returning the query results as an [`sf` object](https://r-spatial.github.io/sf/) in your R session.

Reviewer: ```suggestion
Many geospatial datasets in the B.C. Data Catalogue are available as a [Web Map or Web Feature Service (WMS/WFS)](https://www2.gov.bc.ca/gov/content?id=95D78D544B244F34B89223EF069DF74E). While `bcdc_get_data()` will retrieve the WMS/WFS geospatial data automatically from the web for you (and return it as an `sf` object), sometimes the geospatial file is very large--and slow to download---and you may only want _some_ of the data. `bcdc_query_geodata()` let's you query catalogue geospatial data available as WMS/WFS using `select` and `filter` functions (just like in [`dplyr`](https://dplyr.tidyverse.org/), returning the query results as an [`sf` object](https://r-spatial.github.io/sf/) in your R session.
```
